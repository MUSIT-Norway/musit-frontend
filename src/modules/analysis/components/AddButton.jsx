// @flow

import React from 'react';
import { Col, Button } from 'react-bootstrap';
import FontAwesome from 'react-fontawesome';

type Props = {
  id: string,
  label: string,
  md: number,
  mdOffset?: number
}

const AddButton = ({id, label, md, mdOffset, ...props}: Props) => (
  <div id={id}>
    <Col md={md} mdOffset={mdOffset}>
      <Button {... props}>
        <FontAwesome name='plus-circle'/>{' '}
        {label}
      </Button>
    </Col>
  </div>
);

AddButton.defaultProps = {
  mdOffset: 0
};

export default AddButton;