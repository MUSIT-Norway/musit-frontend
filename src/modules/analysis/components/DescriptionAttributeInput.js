// @flow
import React from 'react';

type Props = {
  attr: {
    attributeKey: string,
    attributeType: string
  },
  onChange: (name: string, type: string) => (e: *) => void,
  value: ?string | ?Array<string | number>
};

const DescriptionAttributeInput = (props: Props) => {
  return (
    <input
      className="form-control"
      type="text"
      name={props.attr.attributeKey}
      onChange={props.onChange(props.attr.attributeKey, props.attr.attributeType)}
      value={(props.value && props.value.toString()) || ''}
    />
  );
};

export default DescriptionAttributeInput;
