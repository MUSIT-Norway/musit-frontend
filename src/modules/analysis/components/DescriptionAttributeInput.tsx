// @flow
import * as React from 'react';
import { ChangeEventHandler } from 'react';
import { DescriptionAttributeType } from './DescriptionAttributeType';
import { Maybe } from '../../../types/common';

interface DescriptionAttributeInputProps {
  attr: DescriptionAttributeType;
  onChange: ChangeEventHandler<HTMLElement>;
  value: Maybe<string> | Maybe<Array<string | number>>;
}

const DescriptionAttributeInput = (props: DescriptionAttributeInputProps) => {
  return (
    <input
      className="form-control"
      type="text"
      name={props.attr.attributeKey}
      onChange={props.onChange}
      value={props.value ? props.value.toString() : ''}
    />
  );
};

export default DescriptionAttributeInput;
