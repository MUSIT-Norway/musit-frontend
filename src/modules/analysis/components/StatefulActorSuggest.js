// @flow
import React from 'react';
import { ActorSuggest } from '../../../components/suggest/ActorSuggest';
import MusitActor from '../../../models/actor';
import type { AppSession } from '../../../types/appSession';
import type { Actor } from 'types/actor';

export type Props = {
  appSession: AppSession,
  onChange: (actorId: ?string) => void,
  value?: ?string
};

export type State = {
  name: ?string
};

export default class StatefulActorSuggest extends React.Component {
  props: Props;
  state: State;

  constructor(props: Props) {
    super(props);
    this.state = { name: null };
  }

  render() {
    return (
      <ActorSuggest
        appSession={this.props.appSession}
        id="restrictions_requester"
        value={this.state.name || this.props.value || ''}
        placeHolder="Find actor"
        onChange={(actor: Actor) => {
          this.props.onChange(MusitActor.getActorId(actor));
          this.setState({ name: actor.fn });
        }}
      />
    );
  }
}
