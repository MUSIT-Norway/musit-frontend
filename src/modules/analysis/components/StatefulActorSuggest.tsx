// @flow
import * as React from 'react';
import { ActorSuggest } from '../../../components/suggest/ActorSuggest';
import MusitActor from '../../../models/actor';
import { AppSession } from '../../../types/appSession';
import { Actor } from '../../../types/actor';
import { I18n } from 'react-i18nify';
import { Maybe } from '../../../types/common';

export type Props = {
  id: string;
  appSession: AppSession;
  onChange: (actorId: Maybe<string>) => void;
  value?: Maybe<string>;
};

export type State = {
  name: Maybe<string>;
};

export default class StatefulActorSuggest extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { name: null };
  }

  render() {
    return (
      <ActorSuggest
        appSession={this.props.appSession}
        id={this.props.id}
        value={this.state.name || this.props.value || ''}
        placeHolder={I18n.t('musit.texts.findActor')}
        onChange={(actor: Actor) => {
          this.props.onChange(MusitActor.getActorId(actor));
          this.setState(() => ({ name: actor.fn }));
        }}
      />
    );
  }
}
