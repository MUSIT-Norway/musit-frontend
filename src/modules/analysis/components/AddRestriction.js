// @flow
import React from 'react';
import DatePicker from '../../../components/DatePicker';
import { DATE_FORMAT_DISPLAY, formatISOString } from '../../../shared/util';
import { I18n } from 'react-i18nify';
import type { AppSession } from '../../../types/appSession';
import type { Restriction } from '../../../types/analysis';
import StatefulActorSuggest from './StatefulActorSuggest';
import { FormElement, FormInput } from '../../../forms/components';

type Props = {
  appSession: AppSession,
  restriction: Restriction,
  updateRestriction: (restriction: Restriction) => void
};

export default function AddRestriction(props: Props) {
  return (
    <div>
      <FormElement
        id="restrictedBy"
        label={I18n.t('musit.analysis.restrictions.restrictionsFor')}
        labelWidth={2}
        elementWidth={10}
      >
        <StatefulActorSuggest
          id="restrictedBy"
          value={props.restriction.requesterName}
          appSession={props.appSession}
          onChange={actorId =>
            props.updateRestriction({
              ...props.restriction,
              requester: actorId
            })}
        />
      </FormElement>
      <FormInput
        id="restrictionCause"
        label={I18n.t('musit.analysis.restrictions.reasonForRestriction')}
        labelWidth={2}
        elementWidth={10}
        value={props.restriction.reason}
        onChange={e =>
          props.updateRestriction({
            ...props.restriction,
            reason: e.target.value
          })}
      />
      <FormInput
        id="restrictionCaseNumbers"
        label={I18n.t('musit.analysis.restrictions.caseNumber')}
        labelWidth={2}
        elementWidth={10}
        value={
          (Array.isArray(props.restriction.caseNumbers) &&
            props.restriction.caseNumbers.join(', ')) ||
          ''
        }
        onChange={e =>
          props.updateRestriction({
            ...props.restriction,
            caseNumbers: e.target.value
              .split(',')
              .map(v => v.trim())
              .filter(v => v)
          })}
      />
      <FormElement
        id="restrictionExpirationEndDate"
        label={I18n.t('musit.analysis.restrictions.endDate')}
        labelWidth={2}
        elementWidth={10}
      >
        <DatePicker
          dateFormat={DATE_FORMAT_DISPLAY}
          value={props.restriction.expirationDate || ''}
          onClear={() =>
            props.updateRestriction({
              ...props.restriction,
              expirationDate: null
            })}
          onChange={selectedDate => {
            if (selectedDate !== 'Invalid date') {
              props.updateRestriction({
                ...props.restriction,
                expirationDate: formatISOString(selectedDate)
              });
            }
          }}
        />
      </FormElement>
    </div>
  );
}
