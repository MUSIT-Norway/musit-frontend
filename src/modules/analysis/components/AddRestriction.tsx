// @flow
import * as React from 'react';
import DatePicker, { getNow } from '../../../components/DatePicker';
import { DATE_FORMAT_DISPLAY, formatISOString } from '../../../shared/util';
import { I18n } from 'react-i18nify';
import { AppSession } from '../../../types/appSession';
import { Restriction } from '../../../types/analysis';
import StatefulActorSuggest from './StatefulActorSuggest';
import { FormElement, FormInput } from '../../../forms/components';
import { MUSTFIX } from '../../../types/common';

type Props = {
  appSession: AppSession;
  restriction: Restriction;
  updateRestriction: (restriction: Restriction) => void;
};

export default function AddRestriction(props: Props) {
  return (
    <div>
      <FormElement
        id="restrictedBy"
        label={I18n.t('musit.analysis.restrictions.restrictionsFor')}
        labelWidth={2}
        elementWidth={10}
      >
        <StatefulActorSuggest
          id="restrictedBy"
          value={props.restriction.requesterName}
          appSession={props.appSession}
          onChange={actorId =>
            props.updateRestriction({
              ...props.restriction,
              requester: actorId
            })
          }
        />
      </FormElement>
      <FormInput
        id="restrictionCause"
        label={I18n.t('musit.analysis.restrictions.reasonForRestriction')}
        labelWidth={2}
        elementWidth={10}
        value={props.restriction.reason}
        onChange={e =>
          props.updateRestriction({
            ...props.restriction,
            reason: e.target.value
          })
        }
      />
      <FormInput
        id="restrictionCaseNumbers"
        label={I18n.t('musit.analysis.restrictions.caseNumber')}
        labelWidth={2}
        elementWidth={10}
        value={
          (Array.isArray(props.restriction.caseNumbers) &&
            props.restriction.caseNumbers.join(', ')) ||
          ''
        }
        onChange={e =>
          props.updateRestriction({
            ...props.restriction,
            caseNumbers: e.target.value
              .split(',')
              .map(v => v.trim())
              .filter(v => v)
          })
        }
      />
      <FormElement
        id="restrictionExpirationEndDate"
        label={I18n.t('musit.analysis.restrictions.endDate')}
        labelWidth={2}
        elementWidth={10}
      >
        <DatePicker
          dateFormat={DATE_FORMAT_DISPLAY}
          value={props.restriction.expirationDate || ''}
          defaultValue={getNow()}
          onClear={() =>
            props.updateRestriction({
              ...props.restriction,
              expirationDate: null
            })
          }
          onChange={(selectedDate: MUSTFIX) => {
            if (selectedDate !== 'Invalid date') {
              //TODO: Fix this schizophrenic "does it have this string value? use it as a date later in the same function" code.
              props.updateRestriction({
                ...props.restriction,
                expirationDate: formatISOString(selectedDate)
              });
            }
          }}
        />
      </FormElement>
    </div>
  );
}
