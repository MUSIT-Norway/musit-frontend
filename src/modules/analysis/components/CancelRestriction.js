// @flow
import React from 'react';
import { I18n } from 'react-i18nify';
import type { AppSession } from '../../../types/appSession';
import type { Restriction } from '../../../types/analysis';
import StatefulActorSuggest from './StatefulActorSuggest';
import { FormElement } from '../../../forms/components';

type Props = {
  appSession: AppSession,
  restriction: Restriction,
  updateRestriction: (restriction: Restriction) => void,
  clickCancel: () => void,
  isRestrictionValidForCancellation: boolean
};

export default function CancelRestriction(props: Props) {
  return (
    <div>
      <FormElement
        id="restrictedBy"
        label={I18n.t('musit.analysis.restrictions.cancelledBy')}
        labelWidth={2}
        elementWidth={10}
        hasError={!props.isRestrictionValidForCancellation}
      >
        <StatefulActorSuggest
          id="restrictedBy"
          value={props.restriction.cancelledByName}
          appSession={props.appSession}
          onChange={actorId =>
            props.updateRestriction({
              ...props.restriction,
              cancelledBy: actorId
            })
          }
        />
      </FormElement>
      <FormElement
        id="cancelCause"
        label={I18n.t('musit.analysis.restrictions.reasonForCancelling')}
        labelWidth={2}
        elementWidth={10}
        hasError={!props.isRestrictionValidForCancellation}
      >
        <input
          className="form-control"
          id="cancelCause"
          value={props.restriction.cancelledReason || ''}
          onChange={e =>
            props.updateRestriction({
              ...props.restriction,
              cancelledReason: e.target.value
            })
          }
        />
        {!props.restriction.cancelledStamp && (
          <button
            className="btn btn-default"
            disabled={!props.isRestrictionValidForCancellation}
            onClick={(e: any) => {
              e.preventDefault();
              props.clickCancel();
            }}
          >
            {I18n.t('musit.analysis.restrictions.cancelRestriction')}
          </button>
        )}
      </FormElement>
    </div>
  );
}
