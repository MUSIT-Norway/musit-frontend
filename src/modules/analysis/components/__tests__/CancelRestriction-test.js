// @flow
import React from 'react';
import CancelRestriction from '../CancelRestriction';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import { appSession } from '../../../../testutils/sampleDataForTest';
import { FormElement } from '../../../../forms/components';
import StatefulActorSuggest from '../../components/StatefulActorSuggest';

declare var describe: any;
declare var it: any;
declare var expect: any;

const restriction = {
  reason: 'Initial reason',
  expirationDate: '2017-08-21T08:17:03+00:00',
  caseNumbers: ['This', 'is', 'not', 'it'],
  requester: 'a50da1c5-9359-4b42-b94a-8faa04ff95cd',
  requesterName: 'dummy user',
  registeredStamp: {
    user: 'a50da1c5-9359-4b42-b94a-8faa04ff95cd',
    date: '2017-08-21T08:17:03+00:00'
  }
};

describe('CancelRestriction', () => {
  it('should render cancel reason and cancelledByName', () => {
    const clickCancel = sinon.spy();
    const updateRestriction = sinon.spy();

    const Comp = shallow(
      <CancelRestriction
        clickCancel={clickCancel}
        appSession={appSession}
        restriction={{
          ...restriction,
          cancelledReason: 'Cancel thizz',
          cancelledStamp: {
            user: 'a50da1c5-9359-4b42-b94a-8faa04ff95cd',
            date: '2017-08-21T08:17:03+00:00'
          },
          cancelledByName: 'Dummy userrr'
        }}
        updateRestriction={updateRestriction}
        isRestrictionValidForCancellation={false}
      />
    );

    expect(Comp.find('button').length).toBe(0);

    expect(shallowToJson(Comp)).toMatchSnapshot();
  });

  it('should call clickCancel when button is clicked', () => {
    const clickCancel = sinon.spy();
    const updateRestriction = sinon.spy();
    const preventDefault = sinon.spy();

    const Comp = shallow(
      <CancelRestriction
        clickCancel={clickCancel}
        appSession={appSession}
        restriction={restriction}
        updateRestriction={updateRestriction}
        isRestrictionValidForCancellation={false}
      />
    );

    Comp.find('button').simulate('click', { preventDefault });
    expect(preventDefault.calledOnce).toBe(true);
    expect(clickCancel.calledOnce).toBe(true);
  });

  it('should call updateRestriction when cancelled by is updated', () => {
    const clickCancel = sinon.spy();
    const updateRestriction = sinon.spy();

    const Comp = shallow(
      <CancelRestriction
        clickCancel={clickCancel}
        appSession={appSession}
        restriction={restriction}
        updateRestriction={updateRestriction}
        isRestrictionValidForCancellation={false}
      />
    );

    Comp.find(StatefulActorSuggest)
      .props()
      .onChange('a50da1c5-9359-4b42-b94a-8faa04ff95cd');
    expect(updateRestriction.calledOnce).toBe(true);
    expect(updateRestriction.getCall(0).args[0]).toEqual({
      ...restriction,
      cancelledBy: 'a50da1c5-9359-4b42-b94a-8faa04ff95cd'
    });
  });

  it('should call updateRestriction when cancelled reason is updated', () => {
    const clickCancel = sinon.spy();
    const updateRestriction = sinon.spy();

    const Comp = shallow(
      <CancelRestriction
        clickCancel={clickCancel}
        appSession={appSession}
        restriction={restriction}
        updateRestriction={updateRestriction}
        isRestrictionValidForCancellation={false}
      />
    );

    Comp.find('input#cancelCause').simulate('change', {
      target: { value: 'Some weird cause' }
    });
    expect(updateRestriction.calledOnce).toBe(true);
    expect(updateRestriction.getCall(0).args[0]).toEqual({
      ...restriction,
      cancelledReason: 'Some weird cause'
    });
  });
});
