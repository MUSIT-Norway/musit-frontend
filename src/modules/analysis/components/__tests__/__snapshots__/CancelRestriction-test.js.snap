// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CancelRestriction should render cancel reason and cancelledByName 1`] = `
<div>
  <FormElement
    elementWidth={10}
    hasError={true}
    id="restrictedBy"
    label="Opphevet av"
    labelWidth={2}
  >
    <StatefulActorSuggest
      appSession={
        Object {
          "accessToken": "45667",
          "actor": Object {
            "fn": "Test",
          },
          "collectionId": "1234",
          "language": Object {
            "isEn": false,
            "isNo": true,
          },
          "museumId": 99,
          "rolesForModules": Object {
            "collectionManagementRead": true,
            "collectionManagementWrite": true,
            "documentArchiveRead": true,
            "documentArchiveWrite": true,
            "storageFacilityAdmin": true,
            "storageFacilityRead": true,
            "storageFacilityWrite": true,
          },
        }
      }
      id="restrictedBy"
      onChange={[Function]}
      value="Dummy userrr"
    />
  </FormElement>
  <FormElement
    elementWidth={10}
    hasError={true}
    id="cancelCause"
    label="Ã…rsak til oppheving"
    labelWidth={2}
  >
    <input
      className="form-control"
      id="cancelCause"
      onChange={[Function]}
      value="Cancel thizz"
    />
  </FormElement>
</div>
`;
